Tutorial/demo on writing a simple multi-stream QUIC client

mattcaswell opened the pull request:
This builds on the existing tutorial on writing a blocking QUIC client to add multi-stream support. We also add a corresponding demo.

There's also a few miscellaneous improvements and cleanups to some of the previous tutorial material.

Finally we update quicserver to be able to interoperate with the multi-stream demo code.

slontis commented:
I am not overly fond of the 'guide' folder name.. Can this be renamed to something more relevant?

mattcaswell commented:
> I am not overly fond of the 'guide' folder name.. Can this be renamed to something more relevant?

I'm open to suggestions.

hlandau commented:
`tutorial` maybe?

hlandau commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,367 @@
+
+Once a stream object is no longer needed it should be freed via a call to
+L<SSL_free(3)>. An application should not call L<SSL_shutdown(3)> on it since
+this is only meaningful for connection level B<SSL> objects.
```
Might be useful to note this will automatically do STOP_SENDING

hlandau commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,367 @@
+different thread and call L<SSL_get0_connection(3)> from within that thread then
+you must be careful to not to call any function that uses the connection object
+at the same time as one of the other threads is also using that connection
+object.
```
We should mention that `SSL_accept_stream` is thread safe.

hlandau reviewed (changes requested):
Otherwise looks good.

mattcaswell commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,367 @@
+TODO(QUIC): What happens if SSL_accept_stream()/SSL_new_stream() is called and
+there is no default stream yet? incoming stream policy suggests that all
+incoming streams are rejected by default....but this only applies after a
+default stream has been created?
```
I made some updates to the man pages to better explain this and removed this comment.

mattcaswell commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,367 @@
+
+TODO(QUIC): What does SSL_get_error() return in the event that the
+peer has reset a stream? If SSL_ERROR_SSL how does an application distinguish
+between a stream reset and a connection level fatal error?
```
I've modified the demo and updated the text to show how to distinguish between connection/stream errors using `SSL_get_stream_read_state()`

mattcaswell commented:
All comments addressed and updates pushed. Please take another look.

hlandau commented on doc/man3/SSL_new_stream.pod:
```diff
@@ -35,6 +35,9 @@ This function can only be used to create stream objects for locally-initiated
 streams. To accept incoming streams initiated by a peer, use
 L<SSL_accept_stream(3)>.
 
+Calling SSL_newt_stream() if there is no default stream already present
```
nit: SSL_new_stream

hlandau reviewed (changes requested):
One more nit, otherwise looks good

mattcaswell commented:
Fixed the nit. Also added a link to the new page from the guide introduction.

Please take another look.

mattcaswell commented:
Ping for second review

t8m commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,379 @@
+L<SSL_set_default_stream_mode(3)> for further details.
+
+It is recommended that new multi-stream applications should not use a default
+stream at all and instead should use a spearate stream B<SSL> object for each
```
Typo: separate

t8m commented on doc/man7/ossl-guide-quic-multi-stream.pod:
```diff
@@ -0,0 +1,379 @@
+using the L<SSL_write_ex(3)>, L<SSL_write(3)>, L<SSL_read_ex(3)> or
+L<SSL_read(3)> functions. See the man pages for further details.
+
+In the event of one of these functions not returning a success code then
+you should call L<SSL_get_error(3)> to find out further details about the error.
```
Consider removing `then` in this sentence

mattcaswell replied:
It makes sense to me to keep it?

t8m reviewed (changes requested):
Great work! Just two nits

mattcaswell commented:
I've rebased to resolve a trivial conflict with master, and fixed the typo spotted by @t8m. I assume @hlandau's approval still holds.

@t8m - please take another look.

t8m commented:
Ouch, needs another rebase.

mattcaswell commented:
> Ouch, needs another rebase.

Grrr. Another trivial conflict. This time in quicserver.c. I'm still assuming approvals hold.

mattcaswell commented:
You've got to be kidding me....another conflict...

mattcaswell commented:
More trivial conflicts. This time caused by the `SSL_set_initial_peer_addr` to `SSL_set1_initial_peer_addr` rename. Rebased again. Again I assume the approvals hold

t8m reviewed (approved):
Still OK, no need to restart the timer.

hlandau reviewed (approved):
OK without restarting timer

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mattcaswell commented:
Pushed. Thanks.
